### PR TITLE
CAMEL-10650: global sslContextParameters

### PR DESCRIPTION
**Pull request by nicolaferraro**
Submitting for a review.
It basically adds global ssl parameters that can be configured using spring-boot properties. There is no ide autocompletion at the moment, bu I'll add it. A sample configuration is contained in the tests e.g. of `camel-undertow-starter`.

To create a global ssl config, I needed to take care of:
- Change each component using SSLContextParameters to lookup the new default if no configuration is given
- Not clash with spring xml configuration of SSLContextParameters, so I avoided putting into the registry an instance of SSLContextParameters, instead I put a GlobalSSLContextParametersSupplier class.
- Make sure that taking a default does not change the behaviour of the component: for components where the presence of a SSLContextParameters in the config turns on the SSL mode, I've put a flag named "useGlobalSslContextParameters" to disable the feature. For some components, the default is false (do not use the global config).

For http/tcp components there is no need of such flag since SSL is enabled by other means (e.g. "https://" in the uri). Some components have their own flag to enable SSL.

### CODE REVIEW

**Comment by janstey** [comment]
Oops I think I just created a conflict for you @nicolaferraro ... Was cleaning up old usage of sslContextParametersRef. I like this idea of having a global SSLContextParameters.

**Comment by nicolaferraro** [comment]
Np ;) I'll fix them later..

File: camel-core/src/main/java/org/apache/camel/util/jsse/GlobalSSLContextParametersSupplier.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2016 Red Hat, Inc.
```
Comment (zregvart) [inline camel-core/src/main/java/org/apache/camel/util/jsse/GlobalSSLContextParametersSupplier.java]: I think you have the wrong comment template here :)

File: components/camel-cxf/src/test/java/org/apache/camel/component/cxf/ssl/SslGlobalTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public static void cleanUp() {
+        //System.clearProperty("cxf.config.file");
+    }
+
+    @BeforeClass
+    public static void startService() {
+        //System.getProperties().put("cxf.config.file", "/org/apache/camel/component/cxf/CxfSslContext.xml");
+        //Greeter implementor = new GreeterImpl();
+        //Endpoint.publish(JAXWS_SERVER_ADDRESS, implementor);
+    }
```
Comment (zregvart) [inline components/camel-cxf/src/test/java/org/apache/camel/component/cxf/ssl/SslGlobalTest.java]: You can probably remove those methods as they contain only commented out code

**Comment by zregvart** [comment]
I wonder if we could have a WithSslParameters interface that has default methods for `getSSLContextParameters` and `setSSLContextParameters` that the component/endpoints can implement and use. And that `getSSLContextParameters` is used by them to get the `SSLContextParameters`. Then we can add global or local `SSLContextParameters` decision to `getSSLContextParameters` method of that interface.
This I think would also be beneficial to add reload and vault support for SSLContextParameters in the future.
Also we could consider adding a getter for `SSLContextParameters` to `CamelContext`, and make it more prominent.

File: camel-core/src/main/java/org/apache/camel/util/jsse/GlobalSSLContextParametersSupplier.java (hunk 1 above)
Comment (nicolaferraro) [inline camel-core/src/main/java/org/apache/camel/util/jsse/GlobalSSLContextParametersSupplier.java]: Damn.. upgrade of IntelliJ.. :)

**Comment by nicolaferraro** [comment]
Sure moving the lookup code in a common getter could be ok, and also implementing custom local logic in the components. I've found that many components use different ways of setting the ssl parameters (component properties, endpoint options, both, ...). 

About the reload and vault support, I'd like to understand more... :)

Thinking about it...

**Comment by zregvart** [comment]
Yeah, the vault could hold certificates/keys the producer/consumer needs, and the point is if this is all in one place then plugging in vault support at one place makes all components auto-support it. ([CAMEL-11030](https://issues.apache.org/jira/browse/CAMEL-11030)).

**Comment by nicolaferraro** [comment]
Some thoughts. The integration of a vault service is really a nice feature..
I like the idea of putting the global SSL config in the Camel context, anyway I think that if I put the `getSSLContextParameters` and `setSSLContextParameters` in a interface, in order to allow Camel to inject those values into the components/endpoints, I'll create problems in cases where a user wants to use custom SSL contexts in different components/endpoints (the `setSSLContextParameters` is usually the main entry point to configure such settings at component level).

Providing a event callback like `onSSLContextParameterChanged` is something useful with authentication tokens or credentials because they change often, while I think it's not appropriate for SSL config since it is usually valid for months/years. A callback like `setGlobalSSLContextParameters` can be also used, and it requires to add another field to each component/endpoint using ssl config...

I think the lookup way (maybe from the camel context, maybe done in the getter rather than in the createEndpoint method) is better for this property. Vault can still setup the global SSLContext variable and it will be still picked up from components in the endpoint creation phase...

Wdyt @zregvart ?

**Comment by zregvart** [comment]
All good 👍  @nicolaferraro I think we should worry about vault support when vault support lands, but I would like to see the a common code to configure `SSLContextParameters` that can be extended later to add vault support, whatever that might be -- and the common lookup + event callback would provide that.
For now I think we should focus on the common lookup, and when we design the vault support focus on the reload functionality.
As a side note, certificate lifetime is getting shorter and shorter, perhaps only in browser scenarios, but the tendency is there. For instance we could have vault support for [Let's encrypt](https://letsencrypt.org/) at some point, those certs are valid for 90 days only, so reload would be quite valuable.

**Comment by nicolaferraro** [comment]
2nd attempt.. :)

**Comment by zregvart** [comment]
👍

**Comment by nicolaferraro** [comment]
I'm going to merge this. I've added the enable/disable flag to all affected component. It defaults to "false" as it's a new feature that needs some test..

I've found a problem with the spring-boot config of "https4". Opening a ticket.

**Pull request closed without merging**